### PR TITLE
feature/ALP-32/verify-user-middleware

### PR DESCRIPTION
**Pull request by RamsesDijkstra**
The Parse() method also includes verfication of the token. I have created a separate function to fetch JWT secret, and have removed a bunch of duplicate code.

### CODE REVIEW

**Review by cnaize — CHANGES_REQUESTED** [review]
- looks like we don't need apiParseJWT(), is it used anywhere?
- ValidateJWT() is good, but not perfect, because we call VerifyUser() for each call for almost all EPs and in your PR it calls GetJWTSecret() each time under the hood. I suggest to add ValidateJWTWithSecret(secret) method and cache the secret inside VerifyUser()

**Review by terbooter — APPROVED** [review]

File: universe/common/api/middleware/verify_user.go [hunk 1]
```diff
@@ -7,27 +7,23 @@ import (
...
 	"go.uber.org/zap"
 
-	"github.com/momentum-xyz/ubercontroller/types/entry"
+	"github.com/momentum-xyz/ubercontroller/types/generic"
 	"github.com/momentum-xyz/ubercontroller/universe"
 	"github.com/momentum-xyz/ubercontroller/universe/common/api"
-	"github.com/momentum-xyz/ubercontroller/utils"
 )
 
+var secretStore = generic.NewSyncMap[string, *[]byte](0)
```
Comment (cnaize) [inline universe/common/api/middleware/verify_user.go]: for me it's a bit complicated
I suggest just load the secret into private variable in VerifyUser() before returning gin.HandlerFunc

**Review by cnaize — CHANGES_REQUESTED** [review]

File: universe/common/api/middleware/verify_user.go [hunk 2]
```diff
@@ -7,27 +7,25 @@ import (
 	"github.com/pkg/errors"
 	"go.uber.org/zap"
 
-	"github.com/momentum-xyz/ubercontroller/types/entry"
-	"github.com/momentum-xyz/ubercontroller/universe"
 	"github.com/momentum-xyz/ubercontroller/universe/common/api"
-	"github.com/momentum-xyz/ubercontroller/utils"
 )
 
+var secret *[]byte = nil
```
Comment (cnaize) [inline universe/common/api/middleware/verify_user.go]: you don't need pointer here, because slice is a pointer itself

Comment (cnaize) [reply universe/common/api/middleware/verify_user.go]: anyway let's move it to VerifyUser(), it's not a big deal to get secret few times while node loading

**Review by cnaize — CHANGES_REQUESTED** [review]

File: universe/common/api/middleware/verify_user.go [hunk 3]
```diff
@@ -7,27 +7,24 @@ import (
...
-			api.AbortRequest(c, http.StatusInternalServerError, "failed_to_get_jwt_key", err, log)
-			return
+	var secret []byte = nil
+	if secret == nil {
+		jwtSecret, err := api.GetJWTSecret()
+		if err != nil {
+			err = errors.WithMessage(err, "failed to fetch jwt secret")
+			log.Error(err)
+			return nil
 		}
```
Comment (cnaize) [inline universe/common/api/middleware/verify_user.go]: please keep only this part, you don't need others

File: universe/common/api/middleware/verify_user.go [hunk 4]
```diff
@@ -7,27 +7,24 @@ import (
...
-		)
-		if !ok || jwtSecret == nil {
-			err := errors.New("Middleware: VerifyUser: failed to get jwt_key attribute")
-			api.AbortRequest(c, http.StatusInternalServerError, "failed_to_get_jwt_key", err, log)
-			return
+	var secret []byte = nil
+	if secret == nil {
+		jwtSecret, err := api.GetJWTSecret()
+		if err != nil {
+			err = errors.WithMessage(err, "failed to fetch jwt secret")
```
Comment (cnaize) [inline universe/common/api/middleware/verify_user.go]: please add "Middleware: VerifyUser: " as error prefix to clearly understand which particular part of application failed

**Review by cnaize — CHANGES_REQUESTED** [review]

File: universe/common/api/middleware/verify_user.go (hunk 3 above)
Comment (cnaize) [reply universe/common/api/middleware/verify_user.go]: just add at the bottom
```
secret := []byte(jwtSecret)
```

**Review by cnaize — APPROVED** [review]

**Pull request merged**